op-node/rollup/driver/driver.go

```diff
@@ -321,6 +364,44 @@ func (s *Driver) eventLoop() {
 	}
 }
 
+// syncSafeHeadFromL2 fetches safe and finalized heads from a remote L2 node
+// and updates the local engine state when using safe-source=l2.
+func (s *Driver) syncSafeHeadFromL2(ctx context.Context) error {
+	s.log.Debug("syncSafeHeadFromL2: fetching finalized/safe from remote L2")
+
+	// Fetch finalized and safe refs from remote L2
+	remoteFinalizedRef, err := s.SyncDeriver.Engine.FetchRemoteL2BlockRef(ctx, eth.Finalized)
+	if err != nil {
+		s.log.Warn("Failed to fetch finalized block from remote L2", "err", err)
+		return err
+	}
+
+	remoteSafeRef, err := s.SyncDeriver.Engine.FetchRemoteL2BlockRef(ctx, eth.Safe)
+	if err != nil {
+		s.log.Warn("Failed to fetch safe block from remote L2", "err", err)
+		return err
+	}
+
+	// Check if remote safe block is on our canonical chain
+	localRef, err := s.SyncDeriver.Engine.L2BlockRefByNumber(ctx, remoteSafeRef.Number)
+	if err != nil || localRef.Hash != remoteSafeRef.Hash {
+		// Either we don't have the block yet, or it's a different hash (reorg needed)
+		s.log.Info("Remote safe block not on canonical chain, advancing unsafe head",
+			"remote_safe", remoteSafeRef.Number,
+			"remote_hash", remoteSafeRef.Hash,
+			"local_err", err,
+			"needs_reorg", err == nil && localRef.Hash != remoteSafeRef.Hash)
+		s.SyncDeriver.Engine.SetUnsafeHead(remoteSafeRef)
```

Comment:
May need a way to check engine is EL syncing, and not do this. For example, inspect EngineController's unsafe head and check that it exists, and preserve it.